### PR TITLE
Add MPPReceiverSet, which includes ExchangeReceiver and CoprocessorReader

### PR DESCRIPTION
**Pull request by windtalker**
Signed-off-by: xufei <xufeixw@mail.ustc.edu.cn>

### What problem does this PR solve?

Issue Number: ref #5095

Problem Summary:

### What is changed and how it works?
- Add MPPReceiverSet, which includes both ExchangeReceiver and CoprocessorReader
- Add `cancelAllReceivers` in MPPTask to cancel all the receivers in current task
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGContext.h [hunk 1]
```diff
@@ -304,11 +309,12 @@ class DAGContext
 
     bool columnsForTestEmpty() { return columns_for_test_map.empty(); }
 
-    const std::unordered_map<String, std::shared_ptr<ExchangeReceiver>> & getMPPExchangeReceiverMap() const;
-    void setMPPExchangeReceiverMap(ExchangeReceiverMapPtr & exchange_receiver_map)
+    ExchangeReceiverPtr getMPPExchangeReceiver(const String & executor_id) const;
+    void setMPPReceiverSet(MPPReceiverSetPtr receiver_set)
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGContext.h:313]:
```suggestion
    void setMPPReceiverSet(const MPPReceiverSetPtr & receiver_set)
```

File: dbms/src/Flash/Coprocessor/DAGContext.h [hunk 2]
```diff
@@ -304,11 +309,12 @@ class DAGContext
...
     {
-        mpp_exchange_receiver_map = exchange_receiver_map;
+        mpp_receiver_set = receiver_set;
     }
+    void addCoprocessorReader(CoprocessorReaderPtr coprocessor_reader);
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGContext.h:317]:
```suggestion
    void addCoprocessorReader(const CoprocessorReaderPtr & coprocessor_reader);
```

File: dbms/src/Flash/Mpp/MPPReceiverSet.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+// limitations under the License.
+
+#include <Flash/Mpp/ExchangeReceiver.h>
+#include <Flash/Mpp/MPPReceiverSet.h>
+
+namespace DB
+{
+void MPPReceiverSet::addExchangeReceiver(const String & executor_id, ExchangeReceiverPtr exchange_receiver)
+{
+    RUNTIME_ASSERT(exchange_receiver_map.find(executor_id) == exchange_receiver_map.end(), log, "Duplicate executor_id in DAGRequest");
```
Comment (SeaRise) [inline dbms/src/Flash/Mpp/MPPReceiverSet.cpp:22]:
```suggestion
    RUNTIME_ASSERT(exchange_receiver_map.find(executor_id) == exchange_receiver_map.end(), log, "Duplicate executor_id : {} in DAGRequest", executor_id);
```

File: dbms/src/Flash/Mpp/MPPReceiverSet.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Mpp/ExchangeReceiver.h>
+#include <Flash/Mpp/MPPReceiverSet.h>
+
+namespace DB
+{
+void MPPReceiverSet::addExchangeReceiver(const String & executor_id, ExchangeReceiverPtr exchange_receiver)
```
Comment (SeaRise) [inline dbms/src/Flash/Mpp/MPPReceiverSet.cpp:20]:
```suggestion
void MPPReceiverSet::addExchangeReceiver(const String & executor_id, const ExchangeReceiverPtr & exchange_receiver)
```

File: dbms/src/Flash/Mpp/MPPReceiverSet.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+{
+    RUNTIME_ASSERT(exchange_receiver_map.find(executor_id) == exchange_receiver_map.end(), log, "Duplicate executor_id in DAGRequest");
+    exchange_receiver_map[executor_id] = exchange_receiver;
+}
+
+void MPPReceiverSet::addCoprocessorReader(CoprocessorReaderPtr coprocessor_reader)
```
Comment (SeaRise) [inline dbms/src/Flash/Mpp/MPPReceiverSet.cpp:26]:
```suggestion
void MPPReceiverSet::addCoprocessorReader(const CoprocessorReaderPtr & coprocessor_reader)
```

File: dbms/src/Flash/Coprocessor/DAGContext.h (hunk 2 above)
Comment (windtalker) [reply dbms/src/Flash/Coprocessor/DAGContext.h]: Done

File: dbms/src/Flash/Mpp/MPPReceiverSet.cpp (hunk 3 above)
Comment (windtalker) [reply dbms/src/Flash/Mpp/MPPReceiverSet.cpp]: Done

File: dbms/src/Flash/Mpp/MPPReceiverSet.cpp (hunk 5 above)
Comment (windtalker) [reply dbms/src/Flash/Mpp/MPPReceiverSet.cpp]: Done

**Review by SeaRise — APPROVED** [review]
LGTM

File: dbms/src/Flash/Mpp/MPPReceiverSet.cpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+void MPPReceiverSet::cancel()
+{
+    for (auto & it : exchange_receiver_map)
+    {
+        it.second->cancel();
+    }
+    for (auto & cop_reader : coprocessor_readers)
+        cop_reader->cancel();
```
Comment (yibin87) [inline dbms/src/Flash/Mpp/MPPReceiverSet.cpp:46]: Do we need to check if exception thrown here?

**Review by yibin87 — APPROVED** [review]
others LGTM

File: dbms/src/Flash/Mpp/MPPReceiverSet.cpp (hunk 6 above)
Comment (windtalker) [reply dbms/src/Flash/Mpp/MPPReceiverSet.cpp:46]: For both exchange receiver and coprocessor read, cancel is just set a flag, it should not throw exception.

**Pull request merged**